src/schemaType/DocumentArrayType.ts

```diff
@@ -60,21 +60,24 @@ class DocumentArrayType extends BaseSchemaType {
 	}
 
 	/** Validate the document array */
-	public async validate(documentList: Document[]): Promise<string[]> {
-		// combining all the validation into one array of promise.all
-		// - validation against the documents in the array will return a single object with 0 to n keys - only those with keys indicate errors;
-		// - iterate the returned object and return the messages from each
-		// - flatten the final results
-
-		return (
-			await Promise.all(
-				documentList.map(async (document) => {
-					const documentErrors = await document.validate();
-
-					return Array.from(documentErrors.values()).map((err) => ensureArray(err));
-				}),
-			)
-		).flat(2);
+	public async validate(documentList: Document[]): Promise<Map<string, string[]>> {
+		// Create a map to store the errors. The key will be the index of the document in the list prefixed to the key path of the error
+		const errorsMap = new Map<string, string[]>();
+		await Promise.all(
+			documentList.map(async (document, index) => {
+				const documentErrors = await document.validate();
+
+				const indexString = String(index);
+				documentErrors.forEach((messages, keyPath) => {
+					const errorsMapKey = `${indexString}.${keyPath}`;
```

Comment:
The setting of this variable can be moved inside the `if` block below since its only used in that scenario.

Reply:
Good idea. done.